회원가입 기능 개발

seungjoo9514 opened the pull request:
# 작업내용
• 회원가입 후 비동기로 계좌 생성 기능 개발
• 계좌 생성에 실패하게되면 계좌 생성 5번 재 실행 기능 개발
• 계좌 생성에 실패한 회원의 계좌 생성을 새벽 시간에 스프링 스케쥴러로 개발

seungjoo9514 committed:
초기

seungjoo9514 committed:
초기작업 2

flab-true commented on src/main/java/com/coin/user/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+@RestController
+public class UserController {
+
+	@Autowired
```
`@Autowired` 는 안티패턴으로 권장하지 않습니다. 왜 그럴까요?
https://dev.to/felixcoutinho/you-should-stop-using-spring-autowired-p8i

seungjoo9514 replied:
solid의 s와 d를 위반할 가능성이 높고 순환 종속성을 처리할 때 코드를 더 복잡하게 만들 수 있습니다
@Autowired 대신 생성자 주입을 통해 작성해야하는것 같습니다

flab-true commented on src/main/java/com/coin/user/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+	@Autowired
+	private UserService userService;
+	
+	public static AES256Cipher aes256;
```
`static`을 사용한 이유는 무엇인가요? `AES256Cipher` 싱글톤이라고 하셨는데, 여기에서 new 키워드가 존재합니다. 
static의 쓰임과 이번 프로젝트에서 구현한 싱글톤을 설명해 주세요.

seungjoo9514 replied:
이번 프로젝트에서 구현한 싱글톤은 암호화를 구현한 싱글톤 객체입니다 싱글톤으로 구현한 AES256Cipher 객체를 new 키워드 없이 사용이 가능하지만 
싱글톤으로 구현한 AES256Cipher 객체를 한번만 생성 후 추가적으로 생성하지 않기 위해 작성하였지만 제가 static 사용에 미숙했던것 같습니다

flab-true commented on src/main/java/com/coin/user/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+	
+	public static AES256Cipher aes256;
+	
+	Random random = new Random();
```
접근제어자가 없습니다. 소스 코드는 일관되게 작성하세요. 
User Controller에 Random의 의존성은 제외하는게 맞을 것 같다는 생각입니다.

seungjoo9514 replied:
네 일관되게 작성하겠습니다 해당 내용은 제거하였습니다

flab-true replied:
new 키워드가 아닌 Builder 패턴을 숙지해주세요. 
method chaning 를 사용하여 객체를 반환되면 좋겠습니다.

flab-true commented on src/main/java/com/coin/user/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+	Random random = new Random();
+	
+//	@PostMapping("/test")
+	@RequestMapping("/user/addUser")
```
REST APIs 설계 법칙을 공부해 주새요. API Spec 설계가 모두 잘못되어 있습니다.

seungjoo9514 replied:
네 공부하겠습니다 현재는 클래스에 @RequestMapping("/users")을 작성하였고 각 crud에 맞게 @PostMapping,  @PutMapping("/{id}")을 작성하였습니다

flab-true commented on src/main/java/com/coin/user/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+	
+//	@PostMapping("/test")
+	@RequestMapping("/user/addUser")
+	public void addUser() {
```
void로 구성되어 있습니다. HTTP로 호출하면 어떤 값을 응답 받습니까?

seungjoo9514 replied:
메소드 타입이 void인 경우에는 응답을 받지 못하여 해당 응답의 결과가 성공인지 실패인지 알 수 없습니다 테스트 당시에는 응답값을 받지 않고 싶어 void로 작성하였지만 해당 내용 수정하였습니다

flab-true replied:
void인 경우 응답을 받지 못할까요?
Http Client로 직접 호출하셔서 어떤 정보들을 받는지 확인해 보시면 도움이 될 것 같습니다. 
Http Status를 확인해 보세요.

seungjoo9514 replied:
Http Client로 직접 호출하여 메소드 타입이 void인 경우에 Http Status 200을 응답 받았습니다 응답의 결과가 성공인지 실패인지 알 수 있습니다 void가 아닌 경우에는 응답값과 return값도 확인이 가능합니다
Status Code: 200
getAllHeaders: Content-Length: 0
getAllHeaders: Date: Thu, 07 Mar 2024 06:02:36 GMT
getAllHeaders: Keep-Alive: timeout=60
getAllHeaders: Connection: keep-alive

flab-true commented on src/main/java/com/coin/user/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+			String name = "김승주";
+			String handPhone = "01012341234";
+			
+			UserDto user = new UserDto(userId, encodingPassword, address, name, handPhone);
```
아래 코드는 UserDto로 이관해주세요. 
```Java
			String userId = "ksj" + random.nextInt();
			String password = "ksj";
			String encodingPassword = aes256.aesEncode(password);
			String address = "서울시";
			String name = "김승주";
			String handPhone = "01012341234";
```

seungjoo9514 replied:
네 이관하겠습니다

flab-true replied:
new 키워드가 아닌 Builder 패턴을 숙지해주세요. 
method chaning 를 사용하여 객체를 반환되면 좋겠습니다.

flab-true commented on src/main/java/com/coin/user/controller/UserController.java:
```diff
@@ -0,0 +1,51 @@
+			userService.addUser(user);
+			
+		} catch (Exception e) {
+			e.printStackTrace();
```
`e.printStackTrace()`; 사용하면 성능에 매우 안좋습니다. 왜 그럴까요?
Logback을 이용해서 로그를 출력하는 습관을 만드세요.

seungjoo9514 replied:
불필요한 내용을 모두 출력하고 예외를 추적하는데 오버헤드가 발생한다, 내부적으로 system.err를 사용하여 제어가 힘들고 리소스 비용이 비싼편입니다
대신 e.getStackTrace()[0] 처럼 실제 예외가 발생한 곳의 위치만 로그를 발생시키면 예외를 추적하는데 비용이 적게 듭니다.
Logback을 이용해서 로그를 출력하도록 하겠습니다

flab-true replied:
로그는 반드시 필요한 내용만 찍으시고 `exception/throwable`이 있는 경우 로깅문을 파라미터화 하는 방법 등 읽어보세요.
- https://yangbongsoo.gitbook.io/study/undefined/log

flab-true commented on src/main/java/com/coin/user/service/UserService.java:
```diff
@@ -0,0 +1,5 @@
+package com.coin.user.service;
+
+public interface UserService {
+	int addUser(UserDto user) throws Exception;
```
아래 형태로 변경하세요.
```Java
UserDao.addUser(UserDto userDto);
``` 
`throws Exception`을 사용하지 마시고 `unChheck Exception`을 활용해 주세요. `mehtod segnature`가 너무 지저분해요.

seungjoo9514 replied:
throws Exception 제거하겠습니다

flab-true replied:
Spring에서 Exception을 지원하는 방법을 숙지해 주시고, 이를 활용해주세요.
- https://hyperskill.org/learn/step/19855

flab-true commented on src/main/java/com/coin/user/serviceimpl/UserDao.java:
```diff
@@ -0,0 +1,54 @@
+	}
+	
+	public boolean addAccount(int userNumber) throws Exception {
+		System.out.println("계좌 생성 시작");
```
` System.out.pritln("...")` 사용하면 안됩니다. Java 시간에 이야기했던 주제입니다. 왜 그럴까요?

seungjoo9514 replied:
System.out.pritln를 사용했을때는 단순히 값만을 출력해보기 위해 사용하였습니다만 단순 값 출력을 위해서도 로그를 통해 출력하겠습니다
로그는 에러 발생 상황을 기록하고 문제를 진단하기위해 사용하지만 System.out.pritln은 해당 내용이 기록되지않고, 로그의 출력 레벨을 설정할 수 없고, 
System.out.pritln의 내부 구현에 synchronized를 사용함으로 인해 멀티 쓰레드 환경에서 성능 저하의 원인이 됩니다.

flab-true commented on src/main/java/com/coin/user/serviceimpl/UserDao.java:
```diff
@@ -0,0 +1,54 @@
+        	nextVal = jdbcTemplate.queryForObject(sql, Long.class);
+        }
+        
+        return nextVal.intValue();
```
`rtn == 1` 아닌 경우 예외가 아닌 정상적인 응답을 합니다. 
else에 대한 로직을 추가하는게 맞지 않을까요? 혹은 예외를 명시적으로 응답하는게 명확할 것 같은데요. 고민해 주세요.
```Java
 if(rtn == 1) {
        	sql = "SELECT USER_SEQUENCE.CURRVAL FROM DUAL";
                 return  jdbcTemplate.queryForObject(sql, Long.class);
 }
        
throw new ...
```

seungjoo9514 replied:
else에 대한 로직 추가하였습니다

flab-true commented on src/main/java/com/coin/user/serviceimpl/UserServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+	@Retryable(maxAttempts = 5, backoff = @Backoff(delay = 1000))
+	public int addUser(UserDto user) {
+        int rtn = 0;
+        CompletableFuture<Integer> future = CompletableFuture.supplyAsync(() -> {
```
CompletableFuture.supplyAsync(() -> { ...} )사용하는 이유는 무엇인가요?
userDao.addUser는 요청에 대한 thread에서 수행하고 userDao.addAccount 작업을 비동기로 수행하면 됩니다. 
로그를 출력하실 때 Thread의  Name을 출력해서 확인해 보세요.

seungjoo9514 replied:
CompletableFuture.supplyAsync( 를 실행하면서 새로운 스레드를 생성하는 것을 확인하였습니다 
현재는 new Thread를 통해 userDao.addAccount 작업을 비동기로 구현하였습니다

flab-true commented on src/main/java/com/coin/user/serviceimpl/UserServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+                try {
+                	boolean retry = userDao.addAccount(0);
+					if(!retry) {
+						throw new Exception();
```
예외로 비지니스 로직을 작성하시면 안됩니다. 어떻게 개선하면 좋을까요?

seungjoo9514 replied:
예외 없이 retry횟수 만큼 for문을 수행하도록 수정하였습니다

flab-true replied:
실패에 대한 Event Queue를 활용하는 방법도 있는데요. 한번 고민해 보시면 좋겠어요.

flab-true commented on src/main/java/com/coin/user/serviceimpl/UserServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+            return null;
+        });
+        
+		return rtn;
```
`return rtn`을 사용하는 곳이 있습니까?

seungjoo9514 replied:
컨트롤러에서 응답 메세지를 화면에 전달할 예정이었습니다

flab-true commented on src/main/java/com/coin/user/serviceimpl/UserServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+	@Autowired
+	private UserDao userDao;
+	
+	int maxRetry = 5;
```
접금 제어자 추가하세요. 변하지 않는 정보라면 불편 변수로 적용하세요.

seungjoo9514 replied:
네 불변 변수로 적용하였습니다

flab-true reviewed (commented):
간단한 코멘트 남겼습니다.

seungjoo9514 committed:
수정사항 커밋
- @Autowired 안티패턴으로 해당 내용 제거 > 생성자 주입을 통해 작성
- UserDto
- e.printStackTrace(); 내용 제거 > logback 이용 로그 출력
- throws Exception 제거
- System.out.pritln("") 제거
- 계좌번호 생성 비동기 로직 수정
- 예외에 비지니스 로직 작성한 부분 내용 수정

flab-true reviewed (approved):
다음 수업을 위하여 PR은 승인합니다. 
시장가 체결에 대한 `Order book` 개념이 있습니다. 관련 자료를 가급적 보지 마시고, 
승주님이 생각하고 계신 매도/매수에 대한 데이터 흐름에 대한 비지니스 로직을 준비해 주시면 감사하겠습니다. 수고 많으셨습니다. 
많이 학습하고 오시면 좋겠어요.

flab-true commented:
PR에 대한 승인 권한이 없네요. 승주님이 승인 및 이후 Feature 개발을 진행해주세요.
